autoprecompiles/src/optimizer.rs

```diff
@@ -44,43 +44,44 @@ pub fn optimize<T: FieldElement>(
         stats_logger.log("exec bus optimization", &machine);
     }
 
-    let mut constraint_system = symbolic_machine_to_constraint_system(machine);
+    let mut constraint_system =
+        JournalingConstraintSystem::from(symbolic_machine_to_constraint_system(machine));
 
     loop {
-        let stats = stats_logger::Stats::from(&constraint_system);
+        let stats = stats_logger::Stats::from(constraint_system.system());
         constraint_system = optimization_loop_iteration(
             constraint_system,
             bus_interaction_handler.clone(),
             degree_bound,
             &mut stats_logger,
             bus_map,
         )?;
-        if stats == stats_logger::Stats::from(&constraint_system) {
-            return Ok(constraint_system_to_symbolic_machine(constraint_system));
+        if stats == stats_logger::Stats::from(constraint_system.system()) {
```

Comment:
Maybe the `AsRef` trait could be useful here to be able to pass `constraint_system` directly